iceoryx_examples/waitset/README.md

```diff
@@ -1,6 +1,586 @@
+
+Every _Trigger_ requires a corresponding class method which returns a boolean
+stating if the _Trigger_ was actually triggered or not. In our case these are
+the two const methods `hasPerformedAction` and `isActivated`.
+```cpp
+    bool hasPerformedAction() const noexcept
+    {
+        return m_hasPerformedAction;
+    }
+
+    bool isActivated() const noexcept
+    {
+        return m_isActivated;
+    }
+```
+
+The method `attachTo` attaches our class to a WaitSet but the user has
+to specify which event they would like to attach. Additionally, they can
+set a `triggerId` and a `callback`.
+
+If the parameter event was set to `PERFORMED_ACTION` we call `acquireTrigger`
+on the waitset which will return an `cxx::expected`. The following parameters
+have to be provided.
+ 
+ 1. The origin of the trigger, e.g. `this`
+ 2. A method of which can be called by the trigger to ask if it was triggered.
+ 3. A method which resets the trigger. Used when the WaitSet goes out of scope. 
+ 4. The id of the trigger.
+ 5. A callback with the signature `void (MyTriggerClass * )`.
+ 
+```cpp
+    iox::cxx::expected<iox::popo::WaitSetError>
+    attachTo(iox::popo::WaitSet& waitset,
+                    const MyTriggerClassEvents event,
+                    const uint64_t triggerId,
+                    const iox::popo::Trigger::Callback<MyTriggerClass> callback) noexcept
+    {
+        switch (event)
+        {
+        case MyTriggerClassEvents::PERFORMED_ACTION:
+        {
+            return waitset
+                .acquireTrigger(this,
+                                {this, &MyTriggerClass::hasPerformedAction},
+                                {this, &MyTriggerClass::invalidateTrigger},
+                                triggerId,
+                                callback)
+                .and_then([this](iox::popo::TriggerHandle& trigger) { 
+                    m_actionTrigger = std::move(trigger); });
+        }
+```
+
+When the parameter event has the value `ACTIVATE` we use the `isActivated` method
+for the trigger.
+```cpp
+        case MyTriggerClassEvents::ACTIVATE:
+        {
+            return waitset
+                .acquireTrigger(this,
+                                {this, &MyTriggerClass::isActivated},
+                                {this, &MyTriggerClass::invalidateTrigger},
+                                triggerId,
+                                callback)
+                .and_then([this](iox::popo::TriggerHandle& trigger) { 
+                    m_activateTrigger = std::move(trigger); });
+        }
+```
+
+The next thing on our checklist is the `invalidateTrigger` method used by the WaitSet
+to reset the _Trigger_ when it goes out of scope. Therefore we look up the
+correct trigger first by calling `isLogicalEqualTo` and then `reset` it.
+```cpp
+    void invalidateTrigger(const uint64_t uniqueTriggerId)
+    {
+        if (m_actionTrigger.getUniqueId() == uniqueTriggerId)
```

Comment:
This will add confusion for the client, especially  if storing those ids for later use. This is an actual id (i.e. unique), the other "id" mentioned above is just a user defined trigger type to distinguish groups of triggers.

Reply:
This is only in the case that you would like to create your own triggerable class which is not the typical use case, more the expert one.

Nevertheless, to distinguish these two id's I called one uniqueId and the other one triggerId but I am open for suggestions and better names.
The implementation requires a unique identifier which cannot be set by the user to identify the TriggerHandle and the corresponding Trigger. Additionally, I think there is no need to store this id since it is always stored directly inside of you TriggerHandle.